### ISSUE TITLE
[JENKINS-50210] Kubernetes agent provisioned regardless of container cap

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#2023]
<p>Hi,<br/>
My teams and I are currently facing an issue with the provisioning of slaves with the kubernetes-plugin using podTemplate that always go higher than the limitation we try to set.</p>

<p>We need to run multiple (tens to hundreds depending on the job) slaves in parallel, that will be doing the same task, and thus we create them with the same pod template.<br/>
Those slaves will run with a heavy CPU workload, so we need to limit the amount of attributed CPU to each of them, but also to set a "container cap" to limit the number of them running at the same time. (because we can not lower the CPU limit too much).</p>

<p>We then decided a Container cap of 7 for our first runs, to see what happens.</p>

<p>That where the problems arose. First, we saw that the cap was not respected strictly, as sometimes, 8 or 9 of them were running at the same time. This is not critical, we can always work with that if we remember it.</p>

<p>Where is the real issue I want to address: <br/>
We figured that after our first batch of ~7 finish, Jenkins will provision, at the same time, a huge amount of slaves, without consideration of the container cap.<br/>
To illustrate that, I recreated a simple scenario to replicate the issue, so I ran in parallel 22 podTemplate that only sleep for a certain time, container cap of 7. What happens was that after the 7 first terminate COMPLETLY (even if 2 or 3 of them finish before that), another batch started, but with the result of having 14 containers running at the same time, and one remaining "in queue". (with more, the queue will be the remaining)</p>

<p>The same compartment arises immediately if we augment the container cap when the build is running.</p>

<p>We would like to know if this was a known issue (I tried to find a similar one here or other places), or if you had an idea of what is happening and why?</p>

<p>I put the scripts that we quickly used to reproduce the issue, and parts of the logs of my Jenkins pod (I cleared it a bit as it was more than 5000 lines of duplicated lines) if it can help you with this.<br/>
And if you need further details, feel free to ask we will try our best to answer asap.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=g_dviniere">g_dviniere</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50210" target="_blank">Kubernetes agent provisioned regardless of container cap</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=csanchez">csanchez</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: kubernetes-plugin
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2018-03-16T14:27:59+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Hi,
My teams and I are currently facing an issue with the provisioning of slaves with the kubernetes-plugin using podTemplate that always go higher than the limitation we try to set.</p>

<p>We need to run multiple (tens to hundreds depending on the job) slaves in parallel, that will be doing the same task, and thus we create them with the same pod template.
Those slaves will run with a heavy CPU workload, so we need to limit the amount of attributed CPU to each of them, but also to set a "container cap" to limit the number of them running at the same time. (because we can not lower the CPU limit too much).</p>

<p>We then decided a Container cap of 7 for our first runs, to see what happens.</p>

<p>That where the problems arose. First, we saw that the cap was not respected strictly, as sometimes, 8 or 9 of them were running at the same time. This is not critical, we can always work with that if we remember it.</p>

<p>Where is the real issue I want to address: 
We figured that after our first batch of ~7 finish, Jenkins will provision, at the same time, a huge amount of slaves, without consideration of the container cap.
To illustrate that, I recreated a simple scenario to replicate the issue, so I ran in parallel 22 podTemplate that only sleep for a certain time, container cap of 7. What happens was that after the 7 first terminate COMPLETLY (even if 2 or 3 of them finish before that), another batch started, but with the result of having 14 containers running at the same time, and one remaining "in queue". (with more, the queue will be the remaining)</p>

<p>The same compartment arises immediately if we augment the container cap when the build is running.</p>

<p>We would like to know if this was a known issue (I tried to find a similar one here or other places), or if you had an idea of what is happening and why?</p>

<p>I put the scripts that we quickly used to reproduce the issue, and parts of the logs of my Jenkins pod (I cleared it a bit as it was more than 5000 lines of duplicated lines) if it can help you with this.
And if you need further details, feel free to ask we will try our best to answer asap.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
kubernetes v1.6.1<br/>
Jenkins: v2.89.4 &amp; v2.89.2<br/>
kubernetes-plugin: v1.3.2 &amp; v1.3.1
```
</details>
<details><summary><i>2 attachments</i></summary>

- [Jenkinsfile](https://issues.jenkins.io/secure/attachment/41848/Jenkinsfile)
- [jenklogs.txt](https://issues.jenkins.io/secure/attachment/41846/jenklogs.txt)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> csanchez:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50210?focusedId=332280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332280">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There's already another issue going into detail of overprovisioning iirc</p>
</pre>
</details>

</ul>
</details>

<p>There's already another issue going into detail of overprovisioning iirc</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> csanchez:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50210?focusedId=332334&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332334">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>See <a href="https://issues.jenkins.io/browse/JENKINS-38260" title="Kubernetes plugin does not respect Container Cap" class="issue-link" data-issue-key="JENKINS-38260"><del>JENKINS-38260</del></a></p>
</pre>
</details>

</ul>
</details>

<p>See <a href="https://issues.jenkins.io/browse/JENKINS-38260" title="Kubernetes plugin does not respect Container Cap" class="issue-link" data-issue-key="JENKINS-38260"><del>JENKINS-38260</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/kubernetes-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-38260%22">JENKINS-38260</a>]</i>
